### DISCUSSION TITLE
randomness in nvim-cmp functionality

### DISCUSSION THREAD

**Discussion by aggonis** [VonHeikemen/lsp-zero.nvim#407] [Q&A] [answered]
hello 👋  i've just started trying to integrate the mighty `LSP` features in my `nvim` setup, and so far i've been trying to replicate some advice found either thru `chatgpt` or online... then i discovered your blog posts which do a wonderful job at trying to get people started with as little as possible. 

to start with, i tried structuring my existing setup `LSP` (`lazy`, `mason`, `mason-lspconfig`, `lspconfig`, `nvim-cmp`, and `cmp-nvim-lsp`) to your `Getting Started`/`Plot Twist` config, and whilst everything works as you'd expect (specifically for my python setup - `basedpyright` and `ruff`), there are instances when i start `nvim` and the autocompletion will simply not work... not even by triggering the keybinding for it to show up. i tried setting the `dependencies` "correctly" in my plugin lua configs, but then again, there's still that random element as to whether the autocompletion will start when i start a new instance of `nvim`.

so far my configs are split into individual `plugin.lua` documents:
`mason.lua`:
```
return {
  'williamboman/mason.nvim',
  opts = {},
}
```
`mason-lspconfig.lua`:
```
return {
  'williamboman/mason-lspconfig.nvim',
  dependencies = {
    'williamboman/mason.nvim',
  },
  opts = {
    ensure_installed = { 'basedpyright', 'ruff' }, -- Ensure these LSPs are installed
    automatic_installation = true, -- Automatically install missing LSPs on startup
  },
}
```
`nvim-lspconfig.lua`:
```
return {
  'neovim/nvim-lspconfig',
  dependencies = {
    'williamboman/mason-lspconfig.nvim',
    'hrsh7th/cmp-nvim-lsp',
    -- 'VonHeikemen/lsp-zero.nvim',
  },
  config = function()
    -- local capabilities = require('cmp_nvim_lsp').default_capabilities()
    local lspconfig = require('lspconfig')

    lspconfig.util.default_config.capabilities = vim.tbl_deep_extend(
      'force',
      lspconfig.util.default_config.capabilities,
      require('cmp_nvim_lsp').default_capabilities()
    )

    lspconfig.basedpyright.setup {
      -- capabilities = capabilities,

      on_attach = function(client, bufnr)
        if client.server_capabilities.inlayHintProvider then
          vim.lsp.inlay_hint.enable(true, { bufnr = bufnr })
        end
      end,

      settings = {
        basedpyright = {
          disableOrganizeImports = true,

          analysis = {
            diagnosticMode = "workspace",
            typeCheckingMode = "off",
            logLevel = "Trace",
              -- useLibraryCodeForTypes = false,
          },
        },
        python = {
          pythonPath = "/opt/homebrew/Caskroom/miniconda/base/envs/<my_env_name>/bin/python",
        },
      }
    }

    require('lspconfig').ruff.setup({})
    --   capabilities = capabilities,
    -- })
  end,
}
```
`nvim-cmp.lua`:
```
return {
  'hrsh7th/nvim-cmp',
  dependencies = {
    'hrsh7th/cmp-nvim-lsp', -- LSP completions
    -- 'hrsh7th/cmp-nvim-lsp-signature-help', -- LSP Signature help
    -- 'hrsh7th/cmp-buffer', -- Buffer completions
    -- 'hrsh7th/cmp-path', -- File path completions
  },
  config = function()
    local cmp = require('cmp')
    -- local cmp_action = require('lsp_zero')

    cmp.setup({
      sources = cmp.config.sources({
        { name = 'nvim_lsp' },
        -- { name = 'nvim_lsp_signature_help' },
        -- { name = 'buffer' },
        -- { name = 'path' },
      }),
      snippet = {
        expand = function(args)
          vim.snippet.expand(args.body)
        end,
      },
      mapping = cmp.mapping.preset.insert({
        -- `Enter` key to confirm completion
        ['<CR>'] = cmp.mapping.confirm({select = false}),

        -- Ctrl+Space to trigger completion menu
        ['<C-Space>'] = cmp.mapping.complete(),

        -- Navigate between snippet placeholder
        -- ['<C-f>'] = cmp_action.vim_snippet_jump_forward(),
        -- ['<C-b>'] = cmp_action.vim_snippet_jump_backward(),

        -- Scroll up and down in the completion documentation
        ['<C-u>'] = cmp.mapping.scroll_docs(-4),
        ['<C-d>'] = cmp.mapping.scroll_docs(4),
      }),
    })
  end
}
```
and finally, `cmp-nvim-lsp.lua`:
```
return {
  'hrsh7th/cmp-nvim-lsp',
  config = function()
    require('cmp').setup({
      sources = {
        name = 'nvim_lsp',
      }
    })
  end,
}
```

i tried looking thru the `lazy` profiling tab for instances where the autocompletion works and doesn't, but i cannot find an exact pattern that would lead to either scenario (e.g. i tried checking whether the ordering of the plugins has an effect, albeit based on the `dependencies` attributes that should always be in the same order), and in both cases, the output i get from `basedpyright` for the completion support is:
```
  completionProvider = {
    completionItem = {
      labelDetailsSupport = true
    },
    resolveProvider = true,
    triggerCharacters = { ".", "[", '"', "'" },
    workDoneProgress = true
  },
```
i'm also running the following `nvim` version:
```
NVIM v0.10.1
Build type: Release
LuaJIT 2.1.1723675123
Run "nvim -V1 -v" for more info
```
i should also mention that the other `basedpyright` features like `hover`, `definition`, `rename`, etc. do work even when the autocompletion doesn't, so that leads me to think there's possibly something very fiddly with the `nvim-cmp` setup on my side or something along those lines, as opposed to a `basedpyright` issue. 

would very much appreciate if you had some insights or hints as to how i go about tracing this randomness in the autocompletion (and apologies for the novel-length question ha!)

**Comment by VonHeikemen** [comment]
The first thing I would do is put a `print("hellooooo")` statement in the `config` function of nvim-cmp. To see if the plugin is loaded.

You can add `lazy = false` property. To make sure lazy.nvim is loading the plugin at startup.

**Reply by aggonis** [reply, marked as answer]
thanks @VonHeikemen. turns out the real culprit was my config setup for `hrsh7th/cmp-nvim-lsp`, specifically the following bit:
```
config = function()
    require('cmp').setup({
      sources = {
        name = 'nvim_lsp',
      }
    })
  end,
```
i misunderstood what the [`README.md`](https://github.com/hrsh7th/cmp-nvim-lsp/blob/main/README.md#setup) was suggesting, the `sources` setup was applicable to the main `nvim-cmp` 🤦‍♂️ oh well... 🫠 

so for those setting up this via `lazy.nvim`, all's needed for the `cmp-nvim-lsp` part is:
```
return {
  'hrsh7th/cmp-nvim-lsp',
}
```

one more question for you if i may. i saw you mentioned in one of your blogposts that `nvim-0.11` adds native autocompletion support, which is indeed what i found on their [roadmap](https://neovim.io/roadmap/) as well, does this mean that `nvim-cmp` will be redundant at that point?
